### ISSUE TITLE
Saving error

### ISSUE THREAD

**Issue by SusanBrown** [cwrc/CWRC-GitWriter#9]
I tried opening one of the sample docs, modified it, and tried to save it. I got a helpful message:
![image](https://user-images.githubusercontent.com/545952/34918878-615c9c90-f928-11e7-85e1-c5b5448eb5d2.png)

It asked me for the name of a repo to save it to, and I gave it the name for a new repo, since it seemed to imply it would create a new repo for me.

I got the following message which is a bit odd: I would have anticipated one field for the new repo name and one field for the document name. The label needs to be clarified at least.
![image](https://user-images.githubusercontent.com/545952/34918886-92fb7096-f928-11e7-828d-68f6fc563f97.png)

I got an error message:
![image](https://user-images.githubusercontent.com/545952/34918889-9d4d9f60-f928-11e7-8293-18a3ecbeb563.png)

I would be useful if the error message could tell me what the problem was so I have some idea of how to fix it.

**Comment by jchartrand** [comment]
> On Jan 14, 2018, at 12:44 PM, SusanBrown <notifications@github.com> wrote:
> 
> I tried opening one of the sample docs, modified it, and tried to save it. I got a helpful message:
>  <https://user-images.githubusercontent.com/545952/34918878-615c9c90-f928-11e7-85e1-c5b5448eb5d2.png>
> It asked me for the name of a repo to save it to, and I gave it the name for a new repo, since it seemed to imply it would create a new repo for me.
> 
> 
Right.
> I got the following message which is a bit odd: I would have anticipated one field for the new repo name and one field for the document name. The label needs to be clarified at least.
> 
I’m confused.  Your description makes it sound like first you were asked for a name, then you were shown the form below.  But as far as I know, there is no intermediate step like you are describing (that asks only for a name).

In any case, I think you are asking why there is no separate name for the document and for the repo.  The reason is that it makes it simpler to always use the same name for the XML document file in the repo:  document.xml.

If you would like to allow the XML file in the repo to have any name (specified by the user), it would complicate both the initial creation, but also the loading of documents.  Where the load is now a one step process (choose the repository) it would be a two step process for every load:

1.  choose the repository
2.  choose the file in the repository

Or I could probably add another ‘config’ file of sorts that is automatically generated and holds metadata specific to the GitWriter, like the name of the document in the repository.    This would of course complicate the code a bit, and you’d run the risk that users would delete the file, not knowing what it is.  Or they’d rename the XML file independently, but I suppose that’s already a problem with naming it document.xml.

Either way, let me know if either option sounds reasonable (double selection or config file).

The ‘hint’ beside the ‘Document Name’ field, as it stands says, “The name to give to the new Github repository that will be created for the document’.

which I’d hoped would explain how it works, but clearly doesn’t.  As always, if you’d like different text, please feel free to suggest different text.


>  <https://user-images.githubusercontent.com/545952/34918886-92fb7096-f928-11e7-828d-68f6fc563f97.png>
> I got an error message:
>  <https://user-images.githubusercontent.com/545952/34918889-9d4d9f60-f928-11e7-8293-18a3ecbeb563.png>
> I would be useful if the error message could tell me what the problem was so I have some idea of how to fix it.
> 
> 
Yes - will do.  I’ll show whatever error is available (http or whatever), but we should probably have some kind of generic message to include with all errors, like:

The following error was returned.  Please contact cwrc@ualberta.ca <mailto:cwrc@ualberta.ca> if you need any help.  

Let me know what would be best.

**Comment by SusanBrown** [comment]
I see about the document name and the repo name. We can leave it as is for now and see what people think. I just think the hint on the field isn't as clear as it could be. Maybe reword it to: The name for the document and also for the new Github repository....

Ideally we should gloss the most common errors and describe what to do, so that we don't get too much stuff coming at the cwrc email.

The major problem though is that saving didn't work for me. Any idea why?

**Comment by jchartrand** [comment]
I will look into it, and try to replicate, but if you could provide a bit more info:

- browser you are using, and whether PC or mac
- which sample document you are saving
- any changes you made to the document before saving it, or any other steps you took
- any errors that are in the ‘console’  If you tell me what browser you are using I can tell you how to see the console, but you can generally get to it from the browser menus.

**Comment by SusanBrown** [comment]
chrome Version 63.0.3239.132 (Official Build) (64-bit)

Mac

sample biography.xml document

Closed it but I just replaced the instruction text in a few of the early name tags

I just tried again and this time it asked me to authorize CWRC-Writer to use my github repo at the start of the session, which it did not do previously when I got this error (and which I assume it should not do every time). 

The console said this:
![image](https://user-images.githubusercontent.com/545952/34920298-948c566c-f93e-11e7-8116-dbeed4ac633b.png)

I then tried it again, didn't get asked to reauthorize CWRC-Writer to use github. The console now says this

![image](https://user-images.githubusercontent.com/545952/34920306-c549b632-f93e-11e7-8431-4439a31f38bb.png)

The only thing I did to the biography.xml document in the latter case was to replace what was there with this: `<NICKNAME>[Testing testing]</NICKNAME>`

**Comment by jchartrand** [comment]
I should have also asked for:

- the url of the gitWriter you are using

It shouldn’t normally ask you to authorize a second time, unless you’ve cleared your cache.  Or, you might have used the prod version of GitWriter in one case, and the dev in another.  Or, you might have used a different browser (e.g., Chrome first, Safari second) — each browser will require it’s own login.

**Comment by SusanBrown** [comment]
http//:cwrc-writer.cwrc.ca<http://cwrc-writer.cwrc.ca>
which took me to the https site and asked me to make an exception
at least once I tried it using https from the start to see if that made a difference, but it didn’t.

I believe I had cleared my cache since I had last used it, so that explains it. There’s no easy way to authorize it permanently, I take it?

On Jan 14, 2018, at 4:06 PM, James Chartrand <notifications@github.com<mailto:notifications@github.com>> wrote:

I should have also asked for:

- the url of the gitWriter you are using

It shouldn’t normally ask you to authorize a second time, unless you’ve cleared your cache. Or, you might have used the prod version of GitWriter in one case, and the dev in another. Or, you might have used a different browser (e.g., Chrome first, Safari second) — each browser will require it’s own login.

**Comment by jchartrand** [comment]
> On Jan 14, 2018, at 5:05 PM, SusanBrown <notifications@github.com> wrote:
> 
> http//:cwrc-writer.cwrc.ca<http://cwrc-writer.cwrc.ca> 
> which took me to the https site and asked me to make an exception 

It should not have taken you to the HTTPS version.  I removed that redirect on Friday.  It doesn’t redirect for me.

> at least once I tried it using https from the start to see if that made a difference, but it didn’t. 
> 
> I believe I had cleared my cache since I had last used it, so that explains it. There’s no easy way to authorize it permanently, I take it? 

Depends what you mean by easy, but the only way I can think of would be to permanently store the user’s Github token on the server, across sessions, which would mean:

- building server side storage for user’s Github tokens, so basically maintaining user accounts on the server (and all the associated baggage that goes along with that)
- CWRC would be responsible for their token on the CWRC server. 

 As it stands, the GitWriter takes a hands-off approach towards the Github authentication as much as possible.

 But, aside from all that, if someone is clearing their cache, they could well want everything removed, including this authorization, and might be annoyed if it wasn’t.

>

**Comment by SusanBrown** [comment]
That is odd. I would swear it was redirecting earlier today, but it doesn’t seem to be doing so now.
Still getting the error.


On Jan 14, 2018, at 6:00 PM, James Chartrand <notifications@github.com<mailto:notifications@github.com>> wrote:



> On Jan 14, 2018, at 5:05 PM, SusanBrown <notifications@github.com<mailto:notifications@github.com>> wrote:
>
> http//:cwrc-writer.cwrc.ca<http://cwrc-writer.cwrc.ca><http://cwrc-writer.cwrc.ca>
> which took me to the https site and asked me to make an exception

It should not have taken you to the HTTPS version. I removed that redirect on Friday. It doesn’t redirect for me.

> at least once I tried it using https from the start to see if that made a difference, but it didn’t.
>
> I believe I had cleared my cache since I had last used it, so that explains it. There’s no easy way to authorize it permanently, I take it?

Depends what you mean by easy, but the only way I can think of would be to permanently store the user’s Github token on the server, across sessions, which would mean:

- building server side storage for user’s Github tokens, so basically maintaining user accounts on the server (and all the associated baggage that goes along with that)
- CWRC would be responsible for their token on the CWRC server.

As it stands, the GitWriter takes a hands-off approach towards the Github authentication as much as possible.

But, aside from all that, if someone is clearing their cache, they could well want everything removed, including this authorization, and might be annoyed if it wasn’t.

>

**Comment by jchartrand** [comment]
The problem is that the GitWriter only handles TEI documents.  We’ve talked before about TEI being the main focus, and throughout the GitWriter development we’ve always only dealt with TEI specifics, for example, with the ‘appInfo’ tag, where we put metadata that identifies the document as ‘gitwriter enabled’, and about which we consulted with James Cummings.  But we never implemented this, or any of the other TEI specifics, e.g., where the annotations are stored, for Orlando or CWRC docs.

I also have a vague memory that there was some reason why we deferred doing this.

We could certainly work all this in now, to allow for CWRC and Orlando docs.

**Comment by SusanBrown** [comment]
Well this is news. If GitWriter is only going to handle TEI documents, then surely we need to remove links, templates, temples, schemas from all other schemas. I can't imagine why they were included in the first place. 

I will create a new ticket for this now. 

I was able to create and save to a new repo successfully using one of the TEI templates, so that is very good. 

@ilovan Please record as a decision in the GProject that GitWriter will support only TEI for the foreseeable future. Please modify the GitWriter Readme and any other documentation specific to it to reflect this decision.

**Comment by ilovan** [comment]
I believe they were kept because there's nothing stopping someone from editing a CWRC or Orlando Document in it, as long as they don't try to save it to a Git Repo. 
A workflow someone may use would be to open a biography template in GitWriter, use edit source to replace the template xml with the xml from a locally stored file, make edits using the GitWriter and when done, replace the content of her local document with the edited XML copied from the "Edit source" popup.

I think a better option than removing all the other templates on the landing page and all the other schema references in the schema dropdown in settings would be to specify when hitting "save" that only TEI docs can be saved in GitHUb and provide suggestions on how to handle the workflow if other document (CWRC, Orlando) was edited. 

Not sure what you mean by links and temples, though.

**Comment by SusanBrown** [comment]
This seems to me too arcane to be worth it given the potential for confusing users, nor does it rely on having templates.

If you really want to save the option to use other schema references I suppose we could, but it seems to me cleanest and least error-prone to remove them.

It would be very nice if gitWriter were able to support people outside of CWRC who wanted to use the Orlando and CWRC schemas, but this is not going to be a common use-case compared to TEI, and I think we should focus on TEI for now and getting the editor really solid.

On Jan 17, 2018, at 9:46 AM, Mihaela Ilovan <notifications@github.com<mailto:notifications@github.com>> wrote:


I believe they were kept because there's nothing stopping someone from editing a CWRC or Orlando Document in it, as long as they don't try to save it to a Git Repo.
A workflow someone may use would be to open a biography template in GitWriter, use edit source to replace the template xml with the xml from a locally stored file, make edits using the GitWriter and when done, replace the content of her local document with the edited XML copied from the "Edit source" popup.

I think a better option than removing all the other templates on the landing page and all the other schema references in the schema dropdown in settings would be to specify when hitting "save" that only TEI docs can be saved in GitHUb and provide suggestions on how to handle the workflow if other document (CWRC, Orlando) was edited.

Not sure what you mean by links and temples, though.
